aldryn_forms/contrib/email_notifications/models.py

```diff
@@ -0,0 +1,187 @@
+        recipient_email = self.get_recipient_email()
+
+        if not recipient_email:
+            message = ugettext('Please provide a recipient.')
+            raise ValidationError(message)
+
+    def get_text_variables(self):
+        return list(self.form.get_form_fields_as_choices())
+
+    def get_recipient_name(self):
+        if self.to_name:
+            # manual name takes precedence over user relationship.
+            name = self.to_name
+        elif self.to_user_id:
+            name = self.to_user.get_full_name()
+        else:
+            name = ''
+        return name
+
+    def get_recipient_email(self):
+        if self.to_email:
+            # manual email takes precedence over user relationship.
+            email = self.to_email
+        elif self.to_user_id:
+            email = self.to_user.email
+        else:
+            email = ''
+        return email
+
+    def get_email_context(self, form):
+        get_template = partial(get_theme_template_name, theme=self.theme)
+
+        context = {
+            'form_plugin': self.form,
+            'form_data': form.get_render_data(),
+            'form_name': self.form.name,
+            'email_notification': self,
+            'email_html_theme': get_template(format='html'),
+            'email_txt_theme': get_template(format='txt'),
+        }
+        return context
+
+    def get_email_kwargs(self, form):
+        text_context = form.get_cleaned_data()
+        text_context['form_name'] = self.form.name
+
+        email_context = self.get_email_context(form)
+        email_context['text_context'] = text_context
+
+        kwargs = {
+            'context': email_context,
+            'language': self.form.language,
+            'template_base': 'aldryn_forms/email_notifications/emails/notification',
+        }
+
+        render = partial(render_text, context=text_context)
+
+        recipient_name = self.get_recipient_name()
+
+        recipient_email = self.get_recipient_email()
+        recipient_email = render(recipient_email)
+
+        if recipient_name:
+            recipient_name = render(recipient_name)
+            recipient_email = formataddr((recipient_name, recipient_email))
+
+        kwargs['recipients'] = [recipient_email]
+
+        if self.from_email:
+            from_email = render(self.from_email)
+
+            if self.from_name:
+                from_name = render(self.from_name)
+                from_email = formataddr((from_name, from_email))
+
+            kwargs['from_email'] = from_email
+        return kwargs
+
+    def prepare_email(self, form):
+        email_kwargs = self.get_email_kwargs(form)
+        return construct_mail(**email_kwargs)
+
+    def render_body_text(self, context):
+        return render_text(self.body_text, context)
+
+    def render_body_html(self, context):
+        return render_text(self.body_html, context)
+
+    def render_subject(self, context):
+        return render_text(self.subject, context)
```

Comment:
You think it's a good idea to have `render_*` functions on model? If I'm not mistaken you're using it only in one place - in template tags. Decision up to you.